Src/FluentAssertions/EventRaisingExtensions.cs

```diff
@@ -13,84 +15,171 @@ namespace FluentAssertions
+                    object sender = @event.Parameters.First();
+                    if (ReferenceEquals(sender, expectedSender))
+                    {
+                        eventsForSender.Add(@event);
+                    }
+                    else
+                    {
+                        otherSenders.Add(sender);
+                    }
                 }
-
-                object actualSender = recordedEvent.Parameters.First();
-                Execute.Assertion
-                    .ForCondition(ReferenceEquals(actualSender, expectedSender))
-                    .FailWith("Expected sender {0}, but found {1}.", expectedSender, actualSender);
             }
 
-            return eventRecorder;
+            Execute.Assertion
+                .ForCondition(eventsForSender.Any())
+                .FailWith("Expected sender {0}, but found {1}.",
+                    () => expectedSender,
+                    () => otherSenders.Distinct());
+
+            return new FilteredEventRecording(eventRecording, eventsForSender);
         }
 
         /// <summary>
-        /// Asserts that at least one occurrence of the event had at least one of the arguments matching a predicate.
+        /// Asserts that at least one occurrence of the events had at least one of the arguments matching a predicate. Returns
+        /// only the events that matched that predicate.
         /// </summary>
-        public static IEventRecorder WithArgs<T>(this IEventRecorder eventRecorder, Expression<Func<T, bool>> predicate)
+        public static IEventRecording WithArgs<T>(this IEventRecording eventRecording, Expression<Func<T, bool>> predicate)
         {
             Guard.ThrowIfArgumentIsNull(predicate, nameof(predicate));
 
             Func<T, bool> compiledPredicate = predicate.Compile();
 
-            if (!eventRecorder.First().Parameters.OfType<T>().Any())
+            bool hasArgumentOfRightType = false;
+            var eventsMatchingPredicate = new List<OccurredEvent>();
+
+            foreach (OccurredEvent @event in eventRecording)
             {
-                throw new ArgumentException("No argument of event " + eventRecorder.EventName + " is of type <" + typeof(T) + ">.");
+                var typedParameters = @event.Parameters.OfType<T>().ToArray();
+                if (typedParameters.Any())
+                {
+                    hasArgumentOfRightType = true;
+                }
+
+                if (typedParameters.Any(parameter => compiledPredicate(parameter)))
+                {
+                    eventsMatchingPredicate.Add(@event);
+                }
             }
 
-            if (eventRecorder.All(recordedEvent => !recordedEvent.Parameters.OfType<T>().Any(parameter => compiledPredicate(parameter))))
+
+            if (!hasArgumentOfRightType)
+            {
+                throw new ArgumentException("No argument of event " + eventRecording.EventName + " is of type <" + typeof(T) + ">.");
```

Comment:
Should this be an assertion fail instead?

Reply:
I'm not sure. It looks like a programming error if you provide the wrong generic type to `WithArgs`